### ISSUE TITLE
[Bug]: milvus makes frequent calls to S3's GetObject, resulting in increased costs

### ISSUE THREAD

**Issue by deer-code** [milvus-io/milvus#27617]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version: 2.1.2
- Deployment mode(standalone or cluster): cluster
- MQ type(rocksmq, pulsar or kafka): pulsar
- SDK version(e.g. pymilvus v2.0.0rc2):pymilvus
- OS(Ubuntu or CentOS): Ubuntu
- CPU/Memory: 32C64G
- GPU: 
- Others:
```

### Current Behavior

frequent calls to S3's GetObject

### Milvus Log

[milvus-log.zip](https://github.com/milvus-io/milvus/files/12856084/milvus-log.zip)

**Comment by deer-code** [comment]
<img width="549" alt="001" src="https://github.com/milvus-io/milvus/assets/11961435/7f7e2d0d-148e-4ebf-8731-1f530a4769f6">
<img width="410" alt="002" src="https://github.com/milvus-io/milvus/assets/11961435/f1af25c6-22ac-47e7-9521-43bb9812b1e2">

2 milvus clusters were built in 2022. on 8/26/2023 there was a sudden increase in cost. One of them was reinstalled on 9/21/2023 and the cost decreased. However, the other one was not reinstalled and the costs continue to be incurred.

**Comment by yanliang567** [comment]
@deer-code a few quick questions:
do you deploy milvus across different region or AZ? 
how do you ensure the increased cost is about API request fee?  which column means the api request fee?
how many times are the API request issued for a month?

/assign @deer-code

**Comment by deer-code** [comment]
1.do you deploy milvus across different region or AZ?
       Yes, milvus across different region deployment for the time being for a number of reasons.
2.how do you ensure the increased cost is about API request fee? which column means the api request fee?
     It has been running stably for nearly a year, and since August 26th, GetObject's Data Transfer Internet (Out) (GB) has suddenly increased by 1,000 times per day to 1,500-2,000 G. During this period, the size of the data in our storage has been maintained at about 50 G, and the total amount of data has remained almost unchanged.(Refer to the chart below)
3.how many times are the API request issued for a month?
There is no detailed access log turned on at s3. currently only a dramatic increase in traffic requesting data can be counted.

![image](https://github.com/milvus-io/milvus/assets/11961435/a82e2a33-98f0-4a0f-aa40-e28a6a54dccc)

**Comment by yanliang567** [comment]
@deer-code 
1. Data Transfer Internet Fee means the data transferring fee, and as far as i know if it is across internet, it is very experience.  So the best practice would be that making the application service and data service in the same region/AZ. 
2. Milvus will call getObject when doing compaction, loading, etc... and then the data will be read from S3 to milvus service. I don't think we changed the API request frequency recently. Did you upgrade milvus version around Aug 26th, from which version to which?

**Comment by deer-code** [comment]
1.Data Transfer Internet Fee means the data transferring fee, and as far as i know if it is across internet, it is very experience. So the best practice would be that making the application service and data service in the same region/AZ.

Thanks for the suggestion. Previously there was cross-region operation because the resources were not ready. Currently switched to the same region.

2.Milvus will call getObject when doing compaction, loading, etc... and then the data will be read from S3 to milvus service. I don't think we changed the API request frequency recently. Did you upgrade milvus version around Aug 26th, from which version to which?

Aug 26th I had to add a new milvus due to production issues. but the version has been consistent with the previous one (2.1.2) and the two milvus share a common s3 but use a different rootPath. the GetObject, while the call occurs, is not in the same order of magnitude as before. Our data volume is only 50G, but GetObject's Data Transfer Internet (Out) volume reaches 2T per day.I'm trying to find out why the problem is happening!

**Comment by deer-code** [comment]
![image](https://github.com/milvus-io/milvus/assets/11961435/1ce07489-686e-41fa-9e39-7a2ec4d669c4)

**Comment by xiaofan-luan** [comment]
> Additional note: our architecture. We currently built two milvus to form a primary and backup model, the primary and backup share s3 but use different rootPaths, relying on scripts to synchronize data daily. on August 26th, the primary had production issues and switched to the backup, the production issues were resolved, but the GetObject cost for the backup increased significantly. Yesterday, after cutting back to the primary, the cost was reduced. The main milvus and s3 are also cross-region, but GetObject traffic is only 1G per day on average, and the cost is less than $1, but the backup milvus is more than 1000G, and the cost reaches $400.

could this be triggered because of your backup script is in one region and S3 bucket is in the other?
I guess milvus didn't to getObject very frequent

**Comment by xiaofan-luan** [comment]
we only do getObject when load data

**Comment by deer-code** [comment]
could this be triggered because of your backup script is in one region and S3 bucket is in the other?

At first, I also suspected that it was the backup script. But after terminating the script for a week, the cost of GetObject has not changed. And we use this same backup model in another production environment and the cost stays very low.

**Comment by deer-code** [comment]
If it's generated by load data, then loading 2T of data of that size per day should cause an outage in the service. With our production requests, we get about a hundred per minute, but the production service has been running normally.

> we only do getObject when load data

**Comment by LoveEachDay** [comment]
Could you share the some metrics for s3 requests?

**Comment by deer-code** [comment]
[milvus-prod-sg-Aug-24-31.xlsx](https://github.com/milvus-io/milvus/files/12867692/milvus-prod-sg-Aug-24-31.xlsx)

aws staff provided information on s3 calls made during the days of August 26 (when the cost exception occurred)

**Comment by LoveEachDay** [comment]
![image](https://github.com/milvus-io/milvus/assets/1573213/1c11f24b-86a1-4a63-a812-263180429202)
Cannot find the api requests count for `GetObject`.

**Comment by deer-code** [comment]
Yes, it was mentioned earlier. Because the access log of s3 is not turned on, all the number of calls cannot be counted, only the flow of calls. Now after milvus reinstallation GetObject cost is back to normal. Can't reproduce the previous problem anymore.
> There is no detailed access log turned on at s3. currently only a dramatic increase in traffic requesting data can be counted.

**Comment by LoveEachDay** [comment]
Is there any data operations during this s3 peak usage period?

**Comment by LoveEachDay** [comment]
Such as large data ingestion?

**Comment by xiaofan-luan** [comment]
could you try 2.2.14? 
2.1.2 is too old and I searched code of milvus, there is no trigger of head object when no load is happedning

**Comment by deer-code** [comment]
> Is there any data operations during this s3 peak usage period?
This bucket is only used by milvus. during the observation period at the beginning of October, we shut down all backup and checking type scripts except for milvus. We didn't do anything manually during this period. But the cost is still more than 400$ per day. Yesterday I reinstalled milvus and only then the charges stopped incurring. That's why I wondered if there was something inside milvus that was causing the constant calls to the GetObject operation.

**Comment by deer-code** [comment]
> could you try 2.2.14? 2.1.2 is too old and I searched code of milvus, there is no trigger of head object when no load is happedning

Thanks for the suggestion, I'm also pushing for an internal project to upgrade the milvus version. And yesterday, the cost of GetObject dropped back down to under $1 after I uninstalled and reinstalled milvus for version 2.1.2.

**Comment by LoveEachDay** [comment]
May be the cost dropping is mainly because the data transfer via internet?

**Comment by deer-code** [comment]
> May be the cost dropping is mainly because the data transfer via internet?

Yes, the cost is lower because the transfer is smaller, only 1G a day. I've been looking for the reason why I was transferring 2T per day before the reinstallation.

**Comment by deer-code** [comment]
<img width="409" alt="image" src="https://github.com/milvus-io/milvus/assets/11961435/03c64680-52db-4696-a547-63558ee2a8b3">

Cost per hour s3. milvus re-installation reduced cost from $18 to $0.90
